### PR TITLE
fix: keep inspect synchronous

### PR DESCRIPTION
**Pull request by gitphill**
- [X] Ready for review
- [X] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Reverts calls to inspect back to synchronous operations to
avoid problems running plugins concurrently.
Keep calls to registry as asynchronous operations.

#### Any background context you want to provide?
After discussing and testing, decided running all monitor operations asynchronously could cause issues (performance, inspects interfering with one another). This change keeps registry calls asynchronous, but changes inspect calls back to synchronous.

### CODE REVIEW

File: src/cli/commands/monitor/index.ts [hunk 1]
```diff
@@ -227,12 +196,74 @@ async function monitor(...args0: MethodArgs): Promise<any> {
...
+      );
+
+      if (inspectResult.plugin.packageManager) {
+        packageManager = inspectResult.plugin.packageManager;
+      }
+
+      analytics.add('pluginName', inspectResult.plugin.name);
+      await spinner.clear(analyzingDepsSpinnerLabel)(inspectResult);
+
+      snapshots.push({
```
Comment (kyegupov) [inline src/cli/commands/monitor/index.ts]: Actually, you don't need to wait for all the inspects to complete to begin sending snapshots. It makes sense to call saveSnapshot here and save the resulting promise. Then, in line 262, you will just need to call Promise.all on that array.

Comment (gitphill) [reply src/cli/commands/monitor/index.ts]: updated, thanks

File: src/cli/commands/monitor/index.ts [hunk 2]
```diff
@@ -227,12 +196,77 @@ async function monitor(...args0: MethodArgs): Promise<any> {
...
+      );
+    } catch (err) {
+      // push this error, the loop continues
+      results.push({ ok: false, data: err, path });
+    }
+  }
+
+  // Part 2: send snapshot to registry asynchronously
+  const snapshotResults = await Promise.all(snapshots);
+  results = results.concat(...snapshotResults);
```
Comment (kyegupov) [inline src/cli/commands/monitor/index.ts]: this .concat means that you don't retain the original order of the arguments. E.g. if you have arguments [a,b,c] and the outcomes are [a=success, b=fail, c=success], the results will be [b-error, a-success, c-success], and they will be printed by the CLI in that order, not mirroring the original.

Comment (gitphill) [reply src/cli/commands/monitor/index.ts]: I couldn't get order to come out like that - but I've updated with another step to ensure snapshots are pushed to results in the order they came in

File: src/cli/commands/monitor/index.ts [hunk 3]
```diff
@@ -227,12 +196,86 @@ async function monitor(...args0: MethodArgs): Promise<any> {
...
+      );
+    } catch (err) {
+      // push this error, the loop continues
+      snapshots.push(Promise.resolve([{ ok: false, data: err, path }]));
+    }
+  }
+
+  // Part 2: send snapshots to registry asynchronously
+  const snapshotResults = await Promise.all(snapshots);
+  const snapshotsResultsFlat = ([] as Array<GoodResult | BadResult>).concat(
```
Comment (kyegupov) [inline src/cli/commands/monitor/index.ts]: is this "concat" necessary? Why can't you just use snapshotResults?

Comment (gitphill) [reply src/cli/commands/monitor/index.ts]: snapshotResults is an array of arrays

Comment (kyegupov) [reply src/cli/commands/monitor/index.ts]: ah, makes sense now. Yes, one argument can produce multiple results (e.g. gradle subprojects).
But then you will lose those extra results when filtering in line 272!

Comment (gitphill) [reply src/cli/commands/monitor/index.ts]: do gradle subprojects have the same path?

Comment (gitphill) [reply src/cli/commands/monitor/index.ts]: ah yea good spot, hmm not sure this approach will be possible then.
the original 

```js
const snapshotResults = await Promise.all(snapshots);
results = ([] as Array<GoodResult | BadResult>).concat(...snapshotResults);
```

seemed to respect the added to snapshots array passed to Promise.all - what about concat could reorder array?

i'll add test for gradle projects

Comment (kyegupov) [reply src/cli/commands/monitor/index.ts]: "seemed to respect the added to snapshots array passed to Promise.all" - I think this is what we should be striving for

**Pull request closed without merging**